Added 'Simulate cross-tab token renew' in test app

denysoblohin-okta opened the pull request:
Internal ref: https://oktainc.atlassian.net/browse/OKTA-456240
Added button 'Simulate cross-tab token renew' in test app that opens 20 iframes that tries to autorenew tokens simultaneously.
Practically system allows to run 15 simultaneous renew requests, next ones will result in HTTP 429.
See video

https://user-images.githubusercontent.com/72614880/153445394-9e862d5c-7ec4-453e-afb7-f293f1b4777f.mov

shuowu commented on test/apps/app/src/testApp.ts:
```diff
@@ -224,6 +237,27 @@ class TestApp {
   }
 
+  async simulateCrossTabTokenRenew(): Promise<void> {
+    const tabsCount = 20; // should be > 15
```
can we add a config in UI for tabs count? so it can be manipulated in e2e

shuowu commented on test/apps/app/src/testApp.ts:
```diff
@@ -224,6 +237,27 @@ class TestApp {
+        `).join('\n');
+        this._setCrossTabsContent(content);
+      } else {
+        alert('Access token is exipred');
```
nit: log instead of alert

denysoblohin-okta commented:
Added config in UI for tabs count and cross-tab sync storage:
<img width="690" alt="Screenshot 2022-02-17 at 12 09 01" src="https://user-images.githubusercontent.com/72614880/154453927-d5ac75bc-2b47-4856-8040-a8902ef386b5.png">

Improved output in iframes to show auth state and renew state (Wating for renew, Expired, Renewed, Updated from other tab) with colors:
<img width="1501" alt="Screenshot 2022-02-17 at 12 08 15" src="https://user-images.githubusercontent.com/72614880/154454848-8e9554ea-5b78-4951-95ee-946fc6302ccd.png">

denysoblohin-okta commented:
Noticed the side effect after  https://github.com/okta/okta-auth-js/pull/1087: there can be situations during cross-tab token renewal when renew fails in one tab with HTTP 429 but can be retried successfully.
See screenshot above - `AuthApiError / Expired / Not authenticated` vs `AuthApiError / Renewed / Authenticated`. In second case there are 2 attempts to renew token.

Explanation:

In `TokenManager.renew()` HTTP 429 error (`AuthApiError`) is caught here:
https://github.com/okta/okta-auth-js/blob/4d8a7b9cb97f53398aa6d765ad33c05b25567b08/lib/TokenManager.ts#L426-L427
Expired token is removed from storage and triggers `EVENT_REMOVED`.
`AuthStateManager` listens to this event and calls `updateAuthState()` here:
https://github.com/okta/okta-auth-js/blob/4d8a7b9cb97f53398aa6d765ad33c05b25567b08/lib/AuthStateManager.ts#L71-L74
In `updateAuthState()` call to `isAuthenticated()` will be performed:
https://github.com/okta/okta-auth-js/blob/4d8a7b9cb97f53398aa6d765ad33c05b25567b08/lib/AuthStateManager.ts#L158
In `isAuthenticated()` token can be renewed if `autoRenew ` options is set:
https://github.com/okta/okta-auth-js/blob/4d8a7b9cb97f53398aa6d765ad33c05b25567b08/lib/OktaAuth.ts#L572-L579

Note two things:
1. we are renewing `accessToken` and `idToken` with one API call (`TokenManager.renew(key)` calls `sdk.token.renewTokens()`)
2. localStorage is shared across tabs

When `TokenManager.renew()` is called for `idToken` and renew fails, ID token is removed from storage - but it could be valid token just renewed from another tab, see (2), we don't check and compare actual token value we're removing.

During `isAuthenticated()` accessToken is checked first for expiration and renew can be triggered here because we've just removed ID token but not Access token, see (1)

Proposals?
- when removing token from storage, check that value is same (has not been updated from another tab)
- ignore `autoRenew` and `autoRemove` in `isAuthenticated()` if we caught error from `renewTokens()`
